### PR TITLE
add voting contract

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: src/contracts/voting.ts [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+            },
+            10
+        )
+
+        for (let i = 0; i < 10; i++) {
+            this.candidates[i] = {
+                name: candidateNames[i],
+                votesReceived: 0n,
+            }
+        }
```
Comment (xhliu) [inline src/contracts/voting.ts:41]: Merge: fill is not needed.

File: src/contracts/voting.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+} from 'scrypt-ts'
+
+export type CandidateName = ByteString
+
+export type Candidate = {
+    name: CandidateName
+    votesReceived: bigint
+}
+
+export type Candidates = FixedArray<Candidate, 10>
```
Comment (xhliu) [inline src/contracts/voting.ts:20]: Change 10 to a constant variable.

File: src/contracts/voting.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    getVotesReceived(candidate: CandidateName): bigint {
+        let votesReceived = 0n
+        for (let i = 0; i < 10; i++) {
+            if (this.candidates[i].name == candidate) {
+                votesReceived = this.candidates[i].votesReceived
+            }
+        }
+
+        return votesReceived
+    }
```
Comment (xhliu) [inline src/contracts/voting.ts:78]: Delete. This should not be a method and trigger a tx. Should be read-only and thus no tx is needed, as in a solidity `call`.

File: tests/local/voting.test.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        let currentInstance = voting
+
+        // call the method of current instance to apply the updates on chain
+        for (let i = 0; i < 10; ++i) {
+            // create the next instance from the current
+            const nextInstance = currentInstance.next()
+
+            const candidate = candidateNames[getRandomInt(0, 10)]
+            // update state
+            nextInstance.updateVotesReceived(candidate)
```
Comment (xhliu) [inline tests/local/voting.test.ts:45]: This method does not exist, right?

File: tests/testnet/voting.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+async function main() {
+    Scrypt.init({
+        apiKey: 'alpha_test_api_key',
+        baseUrl: 'http://localhost:3000',
+    })
+
+    await Voting.compile()
+
+    const hexHash = await Scrypt.contractApi.uploadArtifact(Voting)
```
Comment (xhliu) [inline tests/testnet/voting.ts:14]: Add comments.

File: tests/testnet/voting.ts [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+    console.log('contract Artifact uploaded, hexHash: ', hexHash)
```
Comment (xhliu) [inline tests/testnet/voting.ts:16]: Artifact -> artifact

File: tests/testnet/voting.ts [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        toByteString('candidate1', true),
+        toByteString('candidate2', true),
+        toByteString('candidate3', true),
+        toByteString('candidate4', true),
+        toByteString('candidate5', true),
+        toByteString('candidate6', true),
+        toByteString('candidate7', true),
+        toByteString('candidate8', true),
+        toByteString('candidate9', true),
+        toByteString('candidate10', true),
```
Comment (xhliu) [inline tests/testnet/voting.ts:28]: Refactor: put in a loop

File: tests/testnet/voting.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+                    instance: nextInstance,
+                    balance,
+                },
+            } as MethodCallOptions<Voting>
+        )
+
+        console.log(`Voting call tx: ${tx_i.id}`)
+
+        // update the current instance reference
+        currentInstance = nextInstance
```
Comment (xhliu) [inline tests/testnet/voting.ts:80]: Delete as we agreed on before.

File: tsconfig.json [hunk 9]
```diff
@@ -24,5 +24,5 @@
       }
     ]
   },
-  "include": ["src/**/*.ts", "tests/**/*.ts"]
+  "include": ["src/**/voting.ts", "tests/**/voting.ts"]
```
Comment (xhliu) [inline tsconfig.json:27]: This change shall NOT be committed.

**Review by xhliu — CHANGES_REQUESTED** [review]
👍 

Some improvements.

File: tests/testnet/voting.ts (hunk 7 above)
Comment (zhfnjust) [reply tests/testnet/voting.ts:28]: FixedArray can not be intial with loop

File: tests/testnet/voting.ts [hunk 10]
```diff
@@ -0,0 +1,94 @@
+import { expect } from 'chai'
+import { CandidateName, Voting, N } from '../../src/contracts/voting'
+import {
+    getDefaultSigner,
+    getRandomInt,
+    sleep,
```
Comment (xhliu) [inline tests/testnet/voting.ts:6]: Delete, not used.

File: tests/testnet/voting.ts [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+    stringify,
+} from '../utils/helper'
+import { FixedArray, MethodCallOptions, toByteString, Scrypt } from 'scrypt-ts'
+
+async function main() {
+    Scrypt.init({
+        apiKey: 'alpha_test_api_key',
+        baseUrl:
+            'https://dev-platform-service-testnet-6q7ze.ondigitalocean.app',
```
Comment (xhliu) [inline tests/testnet/voting.ts:15]: - Hide it from user.
- 
- Add a switch for testnet/mainnet

File: tests/local/voting.test.ts (hunk 4 above)
Comment (zhfnjust) [reply tests/local/voting.test.ts]: updateVotesReceived method exists

File: tests/testnet/voting.ts (hunk 11 above)
Comment (zhfnjust) [reply tests/testnet/voting.ts]: we can hide it when connect mainnet:

```ts
    Scrypt.init({
        apiKey: 'alpha_test_api_key',
    })
```

but if connect testnet  or localhost, we need it:

```ts
    Scrypt.init({
        apiKey: 'alpha_test_api_key',
        baseUrl: 'https://testnet.api.scrypt.io',
    })
```

File: tests/testnet/voting.ts (hunk 10 above)
Comment (zhfnjust) [reply tests/testnet/voting.ts]: Deleted

File: tests/testnet/voting.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
+import { CandidateName, Voting, N } from '../../src/contracts/voting'
+import { getDefaultSigner, getRandomInt } from '../utils/helper'
+import { FixedArray, MethodCallOptions, toByteString, Scrypt } from 'scrypt-ts'
+
+async function main() {
+    Scrypt.init({
+        apiKey: 'alpha_test_api_key',
+        baseUrl: 'https://testnet.api.scrypt.io',
```
Comment (xhliu) [inline tests/testnet/voting.ts:8]: Change to network: testnet or mainnet. Uses should not know the url.

File: tests/testnet/voting.ts [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+    TestWallet,
+} from 'scrypt-ts'
+import { myPrivateKey } from '../utils/privateKey'
+
+async function main() {
+    await Voting.compile()
+
+    Scrypt.init({
+        apiKey: 'alpha_test_api_key',
+        network: 'testnet',
```
Comment (xhliu) [inline tests/testnet/voting.ts:18]: Use enum after https://github.com/sCrypt-Inc/scrypt-ts/issues/325

**Pull request merged**